x-pack/metricbeat/module/sql/query/query.go

```diff
@@ -81,167 +79,81 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
-
-	// Extracted from
-	// https://stackoverflow.com/questions/23507531/is-golangs-sql-package-incapable-of-ad-hoc-exploratory-queries/23507765#23507765
-	cols, err := rows.Columns()
-	if err != nil {
-		return errors.Wrap(err, "error getting columns")
-	}
-
-	for k, v := range cols {
-		cols[k] = strings.ToLower(v)
-	}
-
-	vals := make([]interface{}, len(cols))
-	for i := 0; i < len(cols); i++ {
-		vals[i] = new(interface{})
-	}
-
-	for rows.Next() {
-		err = rows.Scan(vals...)
-		if err != nil {
-			m.Logger().Debug(errors.Wrap(err, "error trying to scan rows"))
-			continue
-		}
-
-		numericMetrics := common.MapStr{}
-		stringMetrics := common.MapStr{}
-
-		for i := 0; i < len(vals); i++ {
-			value := getValue(vals[i].(*interface{}))
-			num, err := strconv.ParseFloat(value, 64)
-			if err == nil {
-				numericMetrics[cols[i]] = num
-			} else {
-				stringMetrics[cols[i]] = value
-			}
-
+		for _, ms := range mss {
+			report.Event(m.getEvent(ms))
 		}
 
-		report.Event(mb.Event{
-			RootFields: common.MapStr{
-				"sql": common.MapStr{
-					"driver": m.Driver,
-					"query":  m.Query,
-					"metrics": common.MapStr{
-						"numeric": numericMetrics,
-						"string":  stringMetrics,
-					},
-				},
-			},
-		})
+		return nil
 	}
 
-	if err = rows.Err(); err != nil {
-		m.Logger().Debug(errors.Wrap(err, "error trying to read rows"))
+	ms, err := db.FetchVariableMode(ctx, m.Query)
+	if err != nil {
+		return err
 	}
+	report.Event(m.getEvent(ms))
 
 	return nil
 }
 
-// fetchVariableMode scan the rows and publishes the event for querys that return the response in a key/value format.
-func (m *MetricSet) fetchVariableMode(rows *sqlx.Rows, report mb.ReporterV2) error {
-	data := common.MapStr{}
-	for rows.Next() {
-		var key string
-		var val interface{}
-		err := rows.Scan(&key, &val)
-		if err != nil {
-			m.Logger().Debug(errors.Wrap(err, "error trying to scan rows"))
-			continue
-		}
-
-		key = strings.ToLower(key)
-		data[key] = val
+func (m *MetricSet) getEvent(ms common.MapStr) mb.Event {
+	return mb.Event{
+		RootFields: common.MapStr{
+			"sql": common.MapStr{
+				"driver":  m.Driver,
+				"query":   m.Query,
+				"metrics": getMetrics(ms),
+			},
+		},
 	}
+}
 
-	if err := rows.Err(); err != nil {
-		m.Logger().Debug(errors.Wrap(err, "error trying to read rows"))
-	}
+func getMetrics(ms common.MapStr) (ret common.MapStr) {
+	ret = common.MapStr{}
 
 	numericMetrics := common.MapStr{}
 	stringMetrics := common.MapStr{}
-
-	for key, value := range data {
-		value := getValue(&value)
-		num, err := strconv.ParseFloat(value, 64)
-		if err == nil {
-			numericMetrics[key] = num
-		} else {
-			stringMetrics[key] = value
+	boolMetrics := common.MapStr{}
+
+	for k, v := range ms {
+		switch v.(type) {
+		case float64:
+			numericMetrics.Put(k, v)
+		case string:
+			stringMetrics.Put(k, v)
+		case bool:
+			boolMetrics.Put(k, v)
+		case nil:
+		//Ignore because a nil has no data type and thus cannot be indexed
```

Comment:
Changes here would also require a changelog entry, or to be moved to a different PR.

Reply:
Added